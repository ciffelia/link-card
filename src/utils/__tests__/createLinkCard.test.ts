import { describe, it } from 'vitest';
import {
  createLinkCardFromHtml,
  createLinkCardFromUrl,
} from '../createLinkCard';

describe.concurrent(
  'createLinkCardFromUrl()',
  () => {
    const cases = [
      'http://example.com/',
      'https://ogp.me/',
      'https://www.w3.org/',
      'https://ogp.me/foobarbaz',
      'https://ogp.me/logo.png',
      'http://example.invalid/',
    ] as const satisfies readonly string[];

    for (const url of cases) {
      it(`should return a LinkCard object for ${url}`, async ({ expect }) => {
        expect(await createLinkCardFromUrl(new URL(url))).toMatchSnapshot();
      });
    }
  },
  {
    timeout: 10000,
    retry: 3,
  },
);

describe('createLinkCardFromHtml()', () => {
  const cases = [
    ['', 'https://link-card.example/article/cat'],
    [
      `
          <!doctype html>
          <html>
            <head>
              <link rel="icon" href="/static/favicon.png">
              <meta name="description" content="私はその人を常に先生と呼んでいた。">
              <title>こころ - 夏目漱石</title>
            </head>        
          </html>
        `,
      'https://link-card.example/article/kokoro',
    ],
    [
      `
          <!doctype html>
          <html>
            <head>
              <base href="/article/">
              <link rel="icon" href="images/favicon.png">
              <meta name="description" content="私はその人を常に先生と呼んでいた。">
              <title>こころ - 夏目漱石</title>
              <meta property="og:title" content="こころ">
              <meta property="og:description" content="私はその人を常に先生と呼んでいた。だからここでもただ先生と書くだけで本名は打ち明けない。">
              <meta property="og:image" content="https://link-card.example/static/og.png">
            </head>        
          </html>
        `,
      'https://link-card.example/article/kokoro/detail.html',
    ],
  ] as const satisfies ReadonlyArray<readonly [string, string]>;

  for (const [i, [html, url]] of cases.entries()) {
    it(`should return a LinkCard object #{${i + 1}}`, ({ expect }) => {
      expect(createLinkCardFromHtml(html, new URL(url))).toMatchSnapshot();
    });
  }
});
