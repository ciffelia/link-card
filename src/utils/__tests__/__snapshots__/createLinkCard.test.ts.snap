// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createLinkCardFromHtml() > should return a LinkCard object #{1} 1`] = `
{
  "description": undefined,
  "faviconUrl": "https://link-card.example/favicon.ico",
  "ogImageUrl": undefined,
  "title": undefined,
  "url": "https://link-card.example/article/cat",
}
`;

exports[`createLinkCardFromHtml() > should return a LinkCard object #{2} 1`] = `
{
  "description": "私はその人を常に先生と呼んでいた。",
  "faviconUrl": "https://link-card.example/static/favicon.png",
  "ogImageUrl": undefined,
  "title": "こころ - 夏目漱石",
  "url": "https://link-card.example/article/kokoro",
}
`;

exports[`createLinkCardFromHtml() > should return a LinkCard object #{3} 1`] = `
{
  "description": "私はその人を常に先生と呼んでいた。だからここでもただ先生と書くだけで本名は打ち明けない。",
  "faviconUrl": "https://link-card.example/article/images/favicon.png",
  "ogImageUrl": "https://link-card.example/static/og.png",
  "title": "こころ",
  "url": "https://link-card.example/article/kokoro/detail.html",
}
`;

exports[`createLinkCardFromUrl() > should return a LinkCard object for http://example.com/ 1`] = `
{
  "description": undefined,
  "faviconUrl": "http://example.com/favicon.ico",
  "ogImageUrl": undefined,
  "title": "Example Domain",
  "url": "http://example.com/",
}
`;

exports[`createLinkCardFromUrl() > should return a LinkCard object for http://example.invalid/ 1`] = `
{
  "url": "http://example.invalid/",
}
`;

exports[`createLinkCardFromUrl() > should return a LinkCard object for https://ogp.me/ 1`] = `
{
  "description": "The Open Graph protocol enables any web page to become a rich object in a social graph.",
  "faviconUrl": "https://ogp.me/favicon.ico",
  "ogImageUrl": "https://ogp.me/logo.png",
  "title": "Open Graph protocol",
  "url": "https://ogp.me/",
}
`;

exports[`createLinkCardFromUrl() > should return a LinkCard object for https://ogp.me/foobarbaz 1`] = `
{
  "description": undefined,
  "faviconUrl": "https://ogp.me/favicon.ico",
  "ogImageUrl": undefined,
  "title": undefined,
  "url": "https://ogp.me/foobarbaz",
}
`;

exports[`createLinkCardFromUrl() > should return a LinkCard object for https://ogp.me/logo.png 1`] = `
{
  "description": undefined,
  "faviconUrl": "https://ogp.me/favicon.ico",
  "ogImageUrl": undefined,
  "title": undefined,
  "url": "https://ogp.me/logo.png",
}
`;

exports[`createLinkCardFromUrl() > should return a LinkCard object for https://www.w3.org/ 1`] = `
{
  "description": "The World Wide Web Consortium (W3C) develops standards and guidelines to help everyone build a web based on the principles of accessibility, internationalization, privacy and security.",
  "faviconUrl": "https://www.w3.org/favicon.ico",
  "ogImageUrl": "https://www.w3.org/assets/website-2021/images/w3c-opengraph-image.png",
  "title": "W3C",
  "url": "https://www.w3.org/",
}
`;
